bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/base/restrictions/BasePermissionsExaminationService.java

```diff
@@ -73,40 +74,37 @@ private Collection<AgreementToAccept> agreements(Collection<Requirement> require
 
 	private List<Restriction> assessFeature(Collection<Requirement> requirements, Collection<Permission> permissions,
 			Collection<AgreementToAccept> agreements, Map<Requirement, Permission> active) {
-		List<Restriction> uncovered = insufficientCoverage(requirements, permissions, active);
-		List<Restriction> unaccepted = unacceptedAgreements(agreements);
-		return Arrays.asList(uncovered, unaccepted).stream()//
-				.flatMap(List::stream)//
+		Stream<Restriction> uncovered = insufficientCoverage(requirements, permissions, active);
+		Stream<Restriction> unaccepted = unacceptedAgreements(agreements);
+		return Stream.concat(uncovered, unaccepted) //
 				.collect(Collectors.toList());
 
 	}
 
-	private List<Restriction> insufficientCoverage(Collection<Requirement> requirements,
+	private Stream<Restriction> insufficientCoverage(Collection<Requirement> requirements,
 			Collection<Permission> permissions, Map<Requirement, Permission> active) {
 		return requirements.stream() //
-				.collect(Collectors.groupingBy(Requirement::feature)).entrySet().stream()//
-				.map(e -> insufficientLicenseCoverage(e.getValue(), permissions, active))//
-				.flatMap(Collection::stream)//
-				.collect(Collectors.toList());
+				.collect(Collectors.groupingBy(Requirement::feature)).values().stream()//
+				.flatMap(v -> insufficientLicenseCoverage(v, permissions, active));
 	}
 
-	private Collection<Restriction> insufficientLicenseCoverage(Collection<Requirement> requirements,
+	private Stream<Restriction> insufficientLicenseCoverage(Collection<Requirement> requirements,
 			Collection<Permission> permissions, Map<Requirement, Permission> active) {
 		return requirements.stream()//
 				.filter(requirement -> notCovered(requirement, permissions, active))//
-				.map(requirement -> insufficientLicenseCoverage(requirement)) //
-				.collect(Collectors.toList());
+				.map(this::insufficientLicenseCoverage);
 	}
 
 	private boolean notCovered(Requirement requirement, Collection<Permission> permissions,
 			Map<Requirement, Permission> active) {
-		return !permissions.stream() //
+		Optional<Permission> satisfied = permissions.stream() //
```

Comment:
Became less solid, but more readable and ended up in a comment line removal, which always indicates a right change :)